src/rfq/protocols/bebop/tycho_decoder.rs

```diff
@@ -0,0 +1,228 @@
+use std::collections::HashMap;
+
+use tycho_client::feed::{synchronizer::ComponentWithState};
+use tycho_common::{models::token::Token, Bytes};
+
+use super::{models::BebopPriceData, state::BebopState};
+use crate::{
+    protocol::{errors::InvalidSnapshotError, models::TryFromWithBlock},
+    rfq::models::TimestampHeader,
+};
+
+impl TryFromWithBlock<ComponentWithState, TimestampHeader> for BebopState {
+    type Error = InvalidSnapshotError;
+
+    async fn try_from_with_block(
+        snapshot: ComponentWithState,
+        timestamp_header: TimestampHeader,
+        _account_balances: &HashMap<Bytes, HashMap<Bytes, Bytes>>,
+        all_tokens: &HashMap<Bytes, Token>,
+    ) -> Result<Self, Self::Error> {
+        let state_attrs = snapshot.state.attributes;
+
+        if snapshot.component.tokens.len() != 2 {
+            return Err(InvalidSnapshotError::ValueError(
+                "Component must have 2 tokens (base and quote)".to_string(),
+            ));
+        }
+
+        let base_token_address = snapshot.component.tokens[0].clone();
+        let quote_token_address = snapshot.component.tokens[1].clone();
+
+        let base_token = all_tokens
+            .get(&base_token_address)
+            .ok_or_else(|| {
+                InvalidSnapshotError::ValueError(format!(
+                    "Base token not found: {}",
+                    hex::encode(&base_token_address)
+                ))
+            })?
+            .clone();
+
+        let quote_token = all_tokens
+            .get(&quote_token_address)
+            .ok_or_else(|| {
+                InvalidSnapshotError::ValueError(format!(
+                    "Quote token not found: {}",
+                    hex::encode(&quote_token_address)
+                ))
+            })?
+            .clone();
+
+        let bids_json = state_attrs
+            .get("bids")
+            .ok_or_else(|| InvalidSnapshotError::MissingAttribute("bids".to_string()))?;
+        let asks_json = state_attrs
+            .get("asks")
+            .ok_or_else(|| InvalidSnapshotError::MissingAttribute("asks".to_string()))?;
+
+        // Parse bids and asks from JSON
+        let bids: Vec<(f64, f64)> = serde_json::from_slice(bids_json)
+            .map_err(|e| InvalidSnapshotError::ValueError(format!("Invalid bids JSON: {e}")))?;
+        let asks: Vec<(f64, f64)> = serde_json::from_slice(asks_json)
+            .map_err(|e| InvalidSnapshotError::ValueError(format!("Invalid asks JSON: {e}")))?;
+
+        let price_data =
+            BebopPriceData { last_update_ts: timestamp_header.timestamp as f64, bids, asks };
+
+        Ok(BebopState { base_token, quote_token, price_data })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use tycho_common::{
+        dto::{Chain, ChangeType, ProtocolComponent, ResponseProtocolState},
+        models::Chain as ModelChain,
+    };
+
+    use super::*;
+
+    fn wbtc() -> Token {
```

Comment:
it annoys me that we are repeating this over and over again. cadê o conftest rust?

Reply:
agreeeed